compare_platforms/compare_platforms_overflow.c

```diff
@@ -0,0 +1,134 @@
+{
+    printf("\nOverflowing publicreview by %zx\n", reps);
+    memset(rv->publicreview + offset, 'w', reps);
+    printf("Permissions: %s\n", rv->permissions);
+    fflush(stdout);
+}
+
+char *min(char *a, char *b)
+{
+    return (a > b ? b : a);
+}
+
+char *max(char *a, char *b)
+{
+    return (a > b ? a : b);
+}
+
+void assert_positions(char *publicreview, char *permissions, size_t smsz, size_t bigsz)
+{
+    if (min(publicreview, permissions) == publicreview)
+    {
+        assert((publicreview + bigsz) <= permissions);
+    }
+    else
+    {
+        assert((permissions + smsz) <= publicreview);
+    }
+}
+
+void free_pointers(struct review *rv)
+{
+    /*#if (defined(__x86_64__) && defined(__gnu_linux__))
+        printf("\nOn x86_64 GNU/Linux, you may reach this line, then the attempt to free "
+               "publicreview may cause the program to abort.\n");
+    #endif*/
+
+    free(rv->publicreview);
+    free(rv->permissions);
+}
+
+int main(int argc, char *argv[])
+{
+    struct review review;
+
+    char *publicreview = malloc(0x1001);
+    review.publicreview = strcpy(
+        publicreview, "I am a little unclear as to the contribution. I think the authors could "
+                      "strengthen their case considerably if they conducted an RCT. Weak reject.");
+
+    char *permissions = malloc(0x20);
+    review.permissions = strcpy(permissions, "r");
+
+    print_details(&review);
+
+    assert_positions(publicreview, permissions, 0x20, 0x1001);
+
+    /*#ifdef __CHERI_PURE_CAPABILITY__
+    #if (defined(__aarch64__))
+        printf("\nOn morello-purecap, you may reach this line, then the attempt to overflow by 1 may
+    " "cause a SIGPROT.\n"); #endif #endif*/
+
+    overflow_publicreview(0x1001 + 1, 1, &review);
+
+    const size_t oversz = review.permissions - review.publicreview + 2 - 0x1001;
+
+    /*#if (defined(__arm64__) && defined(__APPLE__))
+        printf("\nOn Apple arm64, you may reach this line, then the attempt to overflow by %zx may "
+               "cause a segmentation fault.\n",
+               oversz);
+    #endif*/
+    /*#if (defined(__x86_64__) && defined(__APPLE__))
+        printf("\nOn Apple x86_64, you may reach this line, then the attempt to overflow by %zx may
+    " "cause a segmentation fault.\n", oversz); #endif*/
+    /*#ifdef __CHERI_PURE_CAPABILITY__
+    #if (defined(__riscv))
+        printf("On riscv64-purecap, you may reach this line, then the attempt to overflow by %zx may
+    " "cause a SIGPROT.\n", oversz); #endif #endif*/
+
+    overflow_publicreview(0x1001 + 2, oversz, &review);
+
+    // If we reached this line, we should have acquired write privileges on the review.
+    print_details(&review);
+
+    free_pointers(&review);
+
+    /*#if (defined(__aarch64__) && defined(__CHERI_CAPABILITY_WIDTH__) && \
+         !defined(__CHERI_PURE_CAPABILITY__))
+        printf("\nOn morello-hybrid, you might make it all the way to the end without crashing.\n");
```

Comment:
Fixed in 4a2cc8a0c1a36a2120ed070d9f1a507486905793. 

I've changed the printfs in the CHERI code blocks from "might" to "should", as each crash is expected behaviour in the respective CHERI environments. I've also added another overflow to the riscv64-purecap conditional block, to show more clearly what the expected behaviour is in that specific environment.

Reply:
I still see a few "may"s in the print messages. I'm just concerned that to me "may" means it could do this, but it's not necessary, or enforced. So I want to make sure we are expecting these crashes to happen when we see these messages, and if they don't, then something went wrong.

Reply:
Fixed in 6df379124a0c3a49f3d8a0b2a87fdea22697e50b.

Reply:
Ok, I'm a bit confused. Did the test before this change have undefined behaviour in non-CHERI environments? Because now this test does not do what I thought it does [1].

[1] https://github.com/capablevms/cheri-examples/pull/82#issuecomment-1424100395

Reply:
Yes, in non-CHERI environments the crash behaviour was undefined. 

The user messages said "may" because they were informational messages, not tests.

Perhaps it would help if I added a single message for non-CHERI environments, like this:

`#if (!defined(__CHERI_CAPABILITY_WIDTH__) && !defined(__CHERI_PURE_CAPABILITY__))`

   ` printf("\nIn a non-CHERI environment, you should reach this line, after which "
    "one of the attempts to overflow may or may not cause a crash.\n ");`

`#endif`

Reply:
Fixed in 96100bc88c984daa920ed9061e6e149609e202e5.